### PR TITLE
Fix transient errors on retry + add more tests

### PR DESCRIPTION
**Pull request by Horneth**
Fix the "non-infinite" retry bug on transient errors

### CODE REVIEW

File: src/test/scala/cromwell/util/TryUtilSpec.scala [hunk 1]
```diff
@@ -72,4 +79,53 @@ class TryUtilSpec extends FlatSpec with Matchers with MockitoSugar {
...
+      retryLimit = Some(5),
+      logger = logger,
+      failMessage = Some(s"failed attempt (on purpose)"),
+      isTransient = isTransient
+    )
+
+    value3.isFailure shouldEqual false
+    value3 shouldEqual Success(9)
+    work3.counter shouldBe 0
+  }
```
Comment (mcovarr) [inline src/test/scala/cromwell/util/TryUtilSpec.scala]: how about a test for the `isFatal` parameter?

**Comment by geoffjentry** [comment]
@Horneth Is this for DSDEEPB-2644?

File: src/test/scala/cromwell/util/TryUtilSpec.scala [hunk 2]
```diff
@@ -72,4 +79,53 @@ class TryUtilSpec extends FlatSpec with Matchers with MockitoSugar {
...
+      backoff = backoff,
+      retryLimit = Some(5),
+      logger = logger,
+      failMessage = Some(s"failed attempt (on purpose)"),
+      isTransient = isTransient
+    )
+
+      value2.isFailure shouldEqual false
+      value2 shouldEqual Success(9)
+      work2.counter shouldBe 0
```
Comment (mcovarr) [inline src/test/scala/cromwell/util/TryUtilSpec.scala]: bad indent

File: src/test/scala/cromwell/util/TryUtilSpec.scala (hunk 1 above)
Comment (Horneth) [inline src/test/scala/cromwell/util/TryUtilSpec.scala]: There is already one :

```
it should "Fail if it hits a fatal exception" in {
     val work = new MockWork
     val value = TryUtil.retryBlock(
       fn = work.failNTimes(4),
       retryLimit = Some(4),
       backoff = backoff,
       logger = logger,
       failMessage = Some(s"failed attempt (on purpose)"),
       isFatal = (t: Throwable) => t.isInstanceOf[IllegalArgumentException]
     )
     value.isFailure shouldEqual true
      value.failed.get.isInstanceOf[CromwellFatalException] shouldBe true
      work.counter shouldBe 3
    }
```

Maybe one that combines `isFatal` + `isTransient` would be good to though.

**Comment by Horneth** [comment]
@geoffjentry yes - I'm updating the board

File: src/test/scala/cromwell/util/TryUtilSpec.scala [hunk 3]
```diff
@@ -13,15 +14,21 @@ import scala.util.Success
 
 class TryUtilSpec extends FlatSpec with Matchers with MockitoSugar {
 
+  class TransientException extends Exception
   class MockWork {
     var counter: Int = _
```
Comment (mcovarr) [inline src/test/scala/cromwell/util/TryUtilSpec.scala]: I hate vars but I'm not going to gate the PR on removing this one.  Though nobody should be surprised if this mysteriously disappears in the near future. :smile:

**Comment by mcovarr** [comment]
:+1:

**Comment by mcovarr** [comment]
Thus Spake Wheel-a-choosa: @scottfrazer

File: src/test/scala/cromwell/util/TryUtilSpec.scala [hunk 4]
```diff
@@ -13,15 +14,21 @@ import scala.util.Success
...
 
-    def failNTimes(n: Int): Option[Int] => Int = {
+    /**
+      * @param n number of times an exception is thrown before succeeding
+      * @param transients how many transients exception to raise (must be <= n)
```
Comment (cjllanwarne) [inline src/test/scala/cromwell/util/TryUtilSpec.scala]: transients exception => transient exceptions

**Comment by scottfrazer** [comment]
:+1:

**Pull request merged**